File: services/auth-service-go/internal/middleware/request_audit.go
package middleware

import (
	"net/http"
	"time"

	chimiddleware "github.com/go-chi/chi/v5/middleware"

	"github.com/oprai/oprai/services/auth-service-go/internal/audit"
	"github.com/oprai/oprai/services/auth-service-go/internal/db"
)

// RequestAudit returns a middleware that logs API errors (4xx/5xx) and slow
// requests (>1s) to the audit_trail table. Successful fast requests are not
// DB-logged to avoid drowning the table with health checks and normal traffic.
func RequestAudit(auditLogger *audit.Logger, trustProxy bool) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Skip health and metrics endpoints — they're noisy and low-value.
			path := r.URL.Path
			if path == "/health" || path == "/metrics" {
				next.ServeHTTP(w, r)
				return
			}

			start := time.Now()
			rec := &statusRecorder{ResponseWriter: w, statusCode: http.StatusOK}
			next.ServeHTTP(rec, r)
			elapsed := time.Since(start)

			status := rec.statusCode
			durationMs := float64(elapsed.Milliseconds())

			// Only emit an audit row for errors or genuinely slow responses.
			if status < 400 && elapsed < time.Second {
				return
			}

			wallet := WalletFromContext(r.Context())

			var eventType, severity string
			switch {
			case status >= 500:
				eventType = "system_error"
				severity = "critical"
			case status == 429:
				eventType = "rate_limit_exceeded"
				severity = "warning"
			case status == 401 || status == 403:
				eventType = "unauthorized_access"
				severity = "error"
			case status >= 400:
				eventType = "api_error"
				severity = "warning"
			default:
				eventType = "api_slow_request"
				severity = "info"
			}

			auditLogger.LogRequest(r, db.AuditEvent{
				EventType:      eventType,
				Severity:       severity,
				WalletAddress:  wallet,
				ResponseStatus: status,
				ResponseTimeMs: durationMs,
				RequestID:      chimiddleware.GetReqID(r.Context()),
			}, trustProxy)
		})
	}
}

File: services/chat-service-py/app/middleware/request_audit.py
"""Request-level audit middleware.

Generates a request_id for every request, binds it to structlog contextvars so
all log lines within the request carry it, and emits an audit DB row for API
errors (4xx/5xx) and slow responses (>1s).
"""

import time
import uuid

import structlog
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response

logger = structlog.get_logger(__name__)


class RequestAuditMiddleware(BaseHTTPMiddleware):
    """Middleware that adds X-Request-ID and emits structured access logs."""

    SKIP_PATHS = {"/health", "/metrics"}

    async def dispatch(self, request: Request, call_next) -> Response:
        # Skip noisy low-value endpoints.
        if request.url.path in self.SKIP_PATHS:
            return await call_next(request)

        request_id = request.headers.get("X-Request-ID") or str(uuid.uuid4())
        wallet = request.headers.get("X-User-Wallet", "")

        # Bind context so all structlog calls within this request include these fields.
        structlog.contextvars.clear_contextvars()
        structlog.contextvars.bind_contextvars(
            request_id=request_id,
            method=request.method,
            path=request.url.path,
            wallet=wallet,
        )

        start = time.perf_counter()
        response = await call_next(request)
        duration_ms = (time.perf_counter() - start) * 1000

        response.headers["X-Request-ID"] = request_id

        status = response.status_code
        log_kwargs = dict(
            status=status,
            duration_ms=round(duration_ms, 2),
        )

        if status >= 500:
            logger.error("request completed", **log_kwargs)
        elif status >= 400:
            logger.warning("request completed", **log_kwargs)
        elif duration_ms > 1000:
            logger.warning("slow request", **log_kwargs)
        else:
            logger.info("request completed", **log_kwargs)

        return response
